client/client.go

```diff
@@ -256,6 +246,116 @@ func (r *NotaryRepository) Initialize(rootKeyIDs []string, serverManagedRoles ..
 	return r.saveMetadata(serverManagesSnapshot)
 }
 
+// certsOfKeyIDs either confirms that the certs and keys (represented by Key IDs) forms valid key pairs.
+// Or throw error when they missmatch.
+// Or generate certificates from the keys if no certificate is provided
+func (r *NotaryRepository) certsOfKeyIDs(keyIDs []string, certs []data.PublicKey) ([]data.PublicKey, error) {
+
+	publicKeys := make([]data.PublicKey, 0, len(keyIDs))
+	// generate certificates if none are provided
+	if certs == nil || len(certs) == 0 {
+		privKeys, err := getAllPrivKeys(keyIDs, r.CryptoService)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, privKey := range privKeys {
+			rootKey, err := rootCertKey(r.gun, privKey)
+			if err != nil {
+				return nil, err
+			}
+			publicKeys = append(publicKeys, rootKey)
+		}
+
+	} else if len(keyIDs) != len(certs) { //return error if length does not match
+		errMsg := fmt.Sprintf("require matching number of keys and certificates but got %d rootkeys and %d certificates", len(keyIDs), len(certs))
+		logrus.Debug(errMsg)
+		return nil, fmt.Errorf(errMsg)
+
+	} else { // match each cert with corresponding Key ID
+		match, err := matchKeyIdsWithCerts(r, keyIDs, certs)
+		if err != nil {
+			return nil, err
+		} else if match {
+			publicKeys = certs
+		}
+	}
+
+	return publicKeys, nil
+}
+
+// matchKeyIdsWithCerts validates that the private keys (represented by their IDs) and the x509 PublicKeys
+// forms matching key pairs
+func matchKeyIdsWithCerts(r *NotaryRepository, ids []string, certs []data.PublicKey) (bool, error) {
```

Comment:
This would be a great opportunity to remove the arbitrary restriction that root public keys be x.509 certs...Perhaps it could be a configuration option on the server side?